File: spark/src/main/java/systematicchaos/studentmobility/GeneralQuery10.java
/**
 * Student Mobility
 * 
 * Javier Fernández-Bravo Peñuela
 * Universitat Politècnica de València
 * 
 * General query 10: Gender proportion of European exchange students.
 * 
 * spark - GeneralQuery10.java
 */

package systematicchaos.studentmobility;

import static org.apache.spark.sql.functions.asc;
import static org.apache.spark.sql.functions.col;
import static systematicchaos.studentmobility.util.Functions.roundHalfDown;

import java.util.ArrayList;
import java.util.List;

import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;

import systematicchaos.studentmobility.domain.Student;
import systematicchaos.studentmobility.spark.SparkJob;
import systematicchaos.studentmobility.spark.SparkJobLauncher;
import systematicchaos.studentmobility.spark.SparkTask;
import systematicchaos.studentmobility.util.KeyCountPair;
import systematicchaos.studentmobility.util.StudentBuilder;

public class GeneralQuery10 {
	
	public static class GenderProportion implements SparkTask<Double> {
		
		@Override
		public Iterable<KeyCountPair<Double>> computeSparkTask(String[] inputFiles) {
			SparkJob<Student> job = new SparkJob<>(GeneralQuery10.class.getName(),
					Student.class, new StudentBuilder());
			List<KeyCountPair<Double>> studentGenderProportion = new ArrayList<>(2);
			
			try {
				Dataset<Student> students = job.readDataset(inputFiles);
				final double studentCount = (double) students.count();
				Dataset<Row> studentsGender = students.groupBy(col("gender")).count();
				
				studentsGender.orderBy(asc("gender")).takeAsList(2).forEach(
						s -> studentGenderProportion.add(new KeyCountPair<>(
								String.valueOf(s.getString(0).charAt(0)),
								roundHalfDown(s.getLong(1) / studentCount))));
			} catch (Exception e) {
				System.err.println(e.getClass().getName() + ": " + e.getMessage());
			} finally {
				job.close();
			}
			
			return studentGenderProportion;
		}
	}
	
	// spark-submit --class systematicchaos.studentmobility.GeneralQuery10 --master local[*] target/spark-1.0.0.jar ../data/Student_Mobility.csv [out/spark10-output]
	public static void main(String[] args) throws Exception {
		SparkJobLauncher.launchSparkJob(new GenderProportion(), args);
	}
}

File: pandas/general_query_10.py
#!/usr/bin/python3

"""General query 10

Student Mobility

Javier Fernández-Bravo Peñuela
Universitat Politècnica de València

Gender proportion of European exchange students.

pandas - general_query_10.py
"""

from pandas_job.pandas_task import compute_task
from domain.student import Student
from util.functions import round_half_down

from sys import argv

def gender_proportion(students_df):
    students_df['GENDER'] = students_df['GENDER'].apply(lambda g: g.value)
    student_count = students_df.shape[0]
    return students_df[['GENDER']].groupby(['GENDER']).size().sort_index(ascending=True)\
        .apply(lambda g: round_half_down(g / student_count))

# python3 general_query_10.py ../data/Student_Mobility.csv [out/pandas10-output]
if __name__ == "__main__":
    compute_task(gender_proportion, Student, argv[1:])

